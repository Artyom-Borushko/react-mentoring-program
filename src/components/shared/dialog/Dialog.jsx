import './dialog.css';
import { createPortal } from 'react-dom';
import React, { useRef } from 'react';
import FocusTrap from 'focus-trap-react';
import useClickOutside from '../../../hooks/ui/useClickOutside';

export default function Dialog({ title, children, onClose }) {
  const modalRef = useRef(null);
  useClickOutside(modalRef, onClose);

  return (
    <>
      {createPortal(
        <FocusTrap
          focusTrapOptions={{
            fallbackFocus: '.fallback-focus',
          }}
        >
          <div className="modal-dialog" ref={modalRef}>
            <img
              src="/images/icons/whiteCrossSign.png"
              alt="white cross"
              className="close-modal-dialog"
              aria-label="close modal dialog"
              onClick={onClose}
              role="presentation"
            />
            <h3
              className="modal-dialog-title"
            >
              {title}
            </h3>
            {children}
          </div>
        </FocusTrap>,
        document.body,
      )}
    </>
  );
}
