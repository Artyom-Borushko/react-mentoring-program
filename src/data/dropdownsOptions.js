export const movieDropdownOptions = ['edit', 'delete'];

export const moviesSortDropdownOptions = ['release date', 'title'];
