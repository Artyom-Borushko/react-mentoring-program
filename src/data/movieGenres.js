
export const listOfGenres = ['all', 'documentary', 'comedy', 'horror', 'crime'];

export const movieGenres = ["Animation", "Adventure", "Family", "Comedy", "Action", "Drama", "Horror", "Sci-Fi"];
