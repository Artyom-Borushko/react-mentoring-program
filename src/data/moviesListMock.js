export const listOfMoviesMock = [
    { src: '/images/movies-list/pulpFiction.png', title: "Pulp Fiction", releaseYear: "2004", relevantGenres: ['action', 'crime'], rating: '9.2', duration: '2h 22min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId1" },
    { src: '/images/movies-list/bohemianRhapsody.png', title: "Bohemian Rhapsody", releaseYear: "2003", relevantGenres: ['documentary', 'drama'], rating: '8.4', duration: '1h 27min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId2" },
    { src: '/images/movies-list/killBillVol2.png', title: "Kill Bill: Vol 2", releaseYear: "1994", relevantGenres: ['comedy', 'crime', 'oscar'], rating: '6.2', duration: '3h 12min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId3" },
    { src: '/images/movies-list/avengersWarOfInfinity.png', title: "Avengers: War of Infinity", releaseYear: "2004", relevantGenres: ['action', 'adventure'], rating: '8.2', duration: '0h 56min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId4" },
    { src: '/images/movies-list/inception.png', title: "Inception", releaseYear: "2003", relevantGenres: ['action', 'adventure', 'horror'], rating: '10.0', duration: '1h 39min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId5" },
    { src: '/images/movies-list/reservoirDogs.png', title: "Reservoir dogs", releaseYear: "1994", relevantGenres: ['crime', 'oscar'], rating: '7.7', duration: '2h 00min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId6" },
];
