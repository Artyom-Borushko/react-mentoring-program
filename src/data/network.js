
export const baseMoviesPath = 'http://localhost:4000/movies';
