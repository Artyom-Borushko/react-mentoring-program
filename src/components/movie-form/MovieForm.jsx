import './movieForm.css';
import React, { useState } from 'react';
import { ErrorMessage, Form, Formik } from 'formik';
import Button from '../shared/button/Button';
import TextInput from '../shared/input/text-input/TextInput';
import Checkbox from '../shared/input/checkbox/Checkbox';
import TextArea from '../shared/input/text-area/TextArea';
import { movieGenres } from '../../data/movieGenres';
import formValidationSchema from '../../schemas/formValidationSchema';

export default function MovieForm({ onSubmit, movie }) {
  const [showDropdown, setShowDropdown] = useState(false);

  const toggleDropdown = () => {
    setShowDropdown(!showDropdown);
  };

  return (
    <Formik
      className="movie-form"
      initialValues={{
        movieTitle: (movie && movie.title) || '',
        movieReleaseDate: (movie && movie.release_date) || '',
        movieUrl: (movie && movie.poster_path) || '',
        movieRating: (movie && movie.vote_average) || '',
        movieGenres: (movie && movie.genres) || [],
        movieRuntime: (movie && movie.runtime) || '',
        movieOverview: (movie && movie.overview) || '',
      }}
      validationSchema={formValidationSchema}
      validateOnBlur={false}
      validateOnChange={false}
      onSubmit={(values, { setSubmitting }) => {
        onSubmit(values);
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, resetForm }) => (
        <Form
          className="movie-form"
        >
          <div className="movie-form-content">
            <TextInput
              label="Title"
              name="movieTitle"
              type="text"
              id="movieTitle"
              placeholder="Title"
            />
            <TextInput
              label="Release date"
              name="movieReleaseDate"
              type="date"
              id="movieReleaseDate"
            />
            <TextInput
              label="Movie URL"
              name="movieUrl"
              type="text"
              id="movieUrl"
              placeholder="https://"
            />
            <TextInput
              label="Rating"
              name="movieRating"
              type="text"
              id="movieRating"
              placeholder="7.8"
            />
            <div className="input-wrapper">
              <button
                type="button"
                className="dropdown-opener"
                onClick={toggleDropdown}
              >
                Select Genres
              </button>
              {showDropdown && (
              <div className="dropdown-content">
                {movieGenres.map((genre) => (
                  <Checkbox
                    key={genre}
                    name="movieGenres"
                    id="movieGenres"
                    value={genre}
                    label={genre}
                  />
                ))}
              </div>
              )}
              <ErrorMessage name="movieGenres">
                {(msg) => <div className="form-field-validation-error">{msg}</div>}
              </ErrorMessage>
            </div>
            <TextInput
              label="Runtime"
              name="movieRuntime"
              type="text"
              id="movieRuntime"
              placeholder="minutes"
            />
            <TextArea
              label="Overview"
              name="movieOverview"
              id="movieOverview"
              placeholder="Movie description"
            />
          </div>

          <div className="movie-form-confirmation-control">
            <Button
              type="reset"
              disabled={isSubmitting}
              onClick={() => resetForm()}
            >
              Reset
            </Button>
            <Button
              type="submit"
              disabled={isSubmitting}
            >
              Submit
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}
