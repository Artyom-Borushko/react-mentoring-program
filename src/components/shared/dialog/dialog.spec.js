import {render, screen} from "@testing-library/react";
import React from "react";
import '@testing-library/jest-dom';
import Dialog from "./Dialog";
import MovieForm from "../../movie-form/MovieForm";
import {listOfMoviesMock} from "../../../data/moviesListMock";

describe("Dialog component tests", () => {

    it("renders empty Dialog component", () => {
        render(<Dialog
            title={'test title'}
        />);

        expect(document.body).not.toHaveTextContent("test123");
    });

    it("renders Dialog component with passed children", () => {
        render(<Dialog
            title={'test title'}
            children={<MovieForm initialMovie={listOfMoviesMock[5]}/>}
        />);

        const input = screen.getByLabelText('Overview');

        expect(input.value).toEqual(listOfMoviesMock[5].description);
    });
});
