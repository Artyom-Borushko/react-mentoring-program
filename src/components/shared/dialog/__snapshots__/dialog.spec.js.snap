// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component tests renders empty Dialog component 1`] = `<div />`;
